src/topp/GNPSExport.cpp

```diff
@@ -79,297 +147,368 @@ using namespace std;
 class TOPPGNPSExport : public TOPPBase
 {
 public:
-  TOPPGNPSExport() :
-  TOPPBase("GNPSExport", "Tool to export consensus features into MGF format", false) {}
+  TOPPGNPSExport() : TOPPBase(
+    "GNPSExport", 
+    "Tool to export representative consensus MS/MS scan per consensusElement into a .MGF file format.\nSee the documentation on https://ccms-ucsd.github.io/GNPSDocumentation/featurebasedmolecularnetworking_with_openms", 
+    true,
+    {
+      { 
+        "Nothias L.F. et al.", // authors
+        "Feature-based Molecular Networking in the GNPS Analysis Environment", // title
+        "bioRxiv 812404 (2019)", // when_where
+        "10.1101/812404" // doi
+      }
+    }
+  ) {}
+
 
 private:
-  double DEF_COSINE_SIMILARITY = 0.95;
-  double DEF_PRECURSOR_MZ_TOLERANCE = 0.0001;
-  double DEF_PRECURSOR_RT_TOLERANCE = 5;
+  static constexpr double DEF_COSINE_SIMILARITY = 0.9;
+  static constexpr double DEF_MERGE_BIN_SIZE = static_cast<double>(BinnedSpectrum::DEFAULT_BIN_WIDTH_HIRES);
+
+  static constexpr double DEF_PREC_MASS_TOL = 0.5;
+  static constexpr bool DEF_PREC_MASS_TOL_ISPPM = false;
+
+  static constexpr double DEF_PEPT_CUTOFF = 5;
+  static constexpr double DEF_MSMAP_CACHE = 50;
+
+
+  void writeMS2BlockToFile(
+    ofstream& output_file,
+    const map<double,int>& ms2_block, 
+    const String& output_type,
+    const int& scan_index,
+    const String& feature_id,
+    const int& feature_charge,
+    const String& feature_mz,
+    const String& spec_index,
+    const String& feature_rt
+  ) 
+  {
+    output_file << "BEGIN IONS" << "\n";
+    output_file << "OUTPUT=" << output_type << "\n";
+    
+    output_file << "SCANS=" << scan_index << "\n";
+    output_file << "FEATURE_ID=e_" << feature_id << "\n";
+
+    output_file << "MSLEVEL=2" << "\n";
+    output_file << "CHARGE=" << to_string(feature_charge == 0 ? 1 : feature_charge) << "+" << "\n";
+    output_file << "PEPMASS=" << feature_mz << "\n";
+    output_file << "FILE_INDEX=" << spec_index << "\n";
+    output_file << "RTINSECONDS=" << feature_rt << "\n";
+
+    for (const pair<double,int>& ms2 : ms2_block)
+    {
+      if (ms2.second > 0)
+      {
+        output_file << ms2.first << "\t" << ms2.second << "\n";
+      }
+    }
+
+    output_file << "END IONS" << "\n\n";
+  }
+
+  void generateMSMSSpectrumBins(
+    const vector<pair<double,int>>& sorted_mz_int_pairs,
+    double delta_mz,
+    map<double,int>& ms2_block
+  )
+  {
+    // generate new spectrum
+    vector<double> mz_merged;
+    vector<double> intensity_merged;
+
+    // double last_mz = numeric_limits<double>::min();
+    double last_mz = sorted_mz_int_pairs[0].first;    
+    double sum_mz = 0;
+    int sum_intensity = 0;
+    double count = 0;
+    for (const auto mz_int : sorted_mz_int_pairs)
```

Comment:
this is probably where the auto-loop copy warning comes from.
use `const auto&` to avoid it